ProjectSidewalk/SidewalkWebpage
# Resolves #593: update config files to support Docker

Resolves #593.

I'm not sure if I was supposed to create a pull request to merge into `develop` or `master` for this repository, but I can update this as needed. We will want these changes on both branches.

### Summary of Changes (in SidewalkWebpage)
* Added an additional resolver for installing dependencies and specified where to get `javax.media` from -- I previously ran into trouble with downloading these because they had been removed from maven.
* Added an environment variable for the Docker database.

### Testing Instructions:

The first half of these instructions are taken from the `README` on the sidewalk-docker respository. Since the last time that we tested, there have been a three main additions to the sidewalk-docker repository: 1) git submodules; 2) caches (speeds up build times because we can pull dependencies locally); 3) auto-deploy scripts for UW servers. These are documented more thoroughly in Jason's pull request [here](https://github.com/aileenzeng/sidewalk-docker/pull/29).

The purpose of these tests are primarily to make sure that we haven't broken anything since previous tests and that the regular dev environment still works properly :)

**Docker Installation**
1. Docker (18.03.1). I am using the Community Edition, version 18.03.1, which can be downloaded [here](https://store.docker.com/search?type=edition&offering=community).
2. Docker Compose (1.21.1). NOTE: Docker for Windows/Mac already come with Docker Compose. Linux users will need to download compose separately from [here]( https://docs.docker.com/compose/install/).

You to see which Docker version you have installed by running the following commands:
```
docker --version
docker-compose --version
```
3. (Windows only) Windows users *must* do the following before cloning: 
    * Give Docker access to drives. Go to Docker > Settings > Shared Drives, and select the drive that you are going to use. This is most likely going to be the C drive.
    * Change git's autocrlf setting. (This must happen before you clone the repository!) Type: `git config --global core.autocrlf false`

### Docker testing
1. Clone the sidewalk-docker repository (Run `git clone https://github.com/aileenzeng/sidewalk-docker.git`.) 
2. In the `resources` folder, place a database dump called `sidewalk.sql` and in the website folder, clone the Project Sidewalk repository. Check out the `593-move-to-docker` branch.
3. Run the following Docker commands to remove any previous containers and volumes: `docker rm -f $(docker ps -a -q)` then `docker volume rm $(docker volume ls -q)`. 
4. Run `docker-compose build`. This step could take a while -- especially the `sbt compile` step.
5. Run `docker-compose up -d db` twice. The first time, it should give a message that says 'Creating sidewalk-docker_db_1 ... done' and the second time, it should say 'Starting sidewalk-docker_db_1 ... done'.
6. Run `docker exec -it sidewalk-docker_db_1 su - postgres -c "createdb -T template0 sidewalk"` to create the database.
7. Run `docker exec -it sidewalk-docker_db_1 su - postgres -c "pg_restore -d sidewalk docker-entrypoint-initdb.d/sidewalk.sql"` to restore the database.
8. Run `docker-compose up`. Something to quickly check:

After step 7, there should be a line that says 
```
website_1  | [info] play - database [default] connected at jdbc:postgresql://db:5432/sidewalk.
```
This is just to make sure that we're connecting to the correct database. 

9. Go to `http://0.0.0.0:9000/` to ensure that everything is working properly. Try auditing a few streets or creating a new user.
10. In a new tab, type `docker exec -it sidewalk-docker_db_1 psql -U sidewalk` to enter the interactive postgres terminal. Do a few select statements that would return information that would have been added in step 8. (Maybe: `select * from sidewalk.user where username like '<new user>';` if you decided to create a new user?). Exit using `\q`.
11. In the tab where you ran the `docker-compose up` command, hit `ctrl-c` to stop the container. Restart it using `docker-compose up`. We're doing this to make sure that the data in our database is persisting between container stops and starts.
12. Repeat Step 9 -- you should be getting the same things if you are running identical queries as before. 
13. After verifying that Sidewalk is working correctly with Docker, hit `ctrl-c` to stop. Type `docker ps`. There shouldn't be any containers that are listed - they should all be stopped.
14. Run Sidewalk normally by starting up vagrant, grunt and sbt.
